accepted/2020/serializer/WriteableDomAndDynamic.md

```diff
@@ -0,0 +1,587 @@
+        public override bool TryGetValue<T>(out T? value);
+        public override bool TryGetValue(Type type, out object? value);
+    }
+
+    // These types are used with dynamic support.
+    // JsonSerializerOptions.EnableDynamicTypes must be called before using.
+    // These types are public for "edit mode" scenarios when using dynamic.
+    public class JsonDynamicObject : JsonObject, IDynamicMetaObjectProvider { }
+    public class JsonDynamicArray : JsonArray, IDynamicMetaObjectProvider { }
+    public class JsonDynamicValue : JsonValue, IDynamicMetaObjectProvider { }
+}
+```
+
+## Serializer interop
+For deserializating JSON to a `JsonNode` instance, the existing static `JsonSerializer.Deserialize()` methods are used.
+
+For serializing a `JsonNode` instance to JSON, instance helpers on `JsonNode` are provided that pass in the value and options:
+```cs
+JsonObject obj = ...
+
+// short version:
+string json = obj.Serialize();
+
+// equivalent longer versions:
+string json = JsonSerializer.Serialize(obj, obj.GetType(), obj.Options);
+string json = JsonSerializer.Serialize<JsonObject>(obj, obj.Options);
+string json = JsonSerializer.Serialize<JsonNode>(obj, obj.Options);
+```
+
+Serializer-specific functionality including custom converters and quoted numbers just work.
+
+The `JsonSerializerOptions` is passed to `JsonNode` and used to specify any run-time added custom converters, handle other options such as quoted numbers, and to respect the `PropertyNameCaseInsensitive` to determine whether the string-based lookup of property names on `JsonObject` are case-sensitive (for both direct use of `JsonObject` and also indirect use through `dynamic`). If we have `static JsonValue Parse()` methods, those would also need an override to specify the `JsonSerializerOptions`.
+
+Although the serializer can be used to directly (de)serialize `JsonNode` types, once a `JsonNode` instance is created, it is expected the `JsonNode.GetValue<T>()` method is used to change types instead of using the serializer:
+```cs
+    JsonNode node = ... // any value
+    string str;
+
+    // This works, but has no caching:
+    byte[] utf8 = node.SerializeToUtf8Bytes();
+    str = JsonSerializer.Deserialize<string>(utf8, options);
+
+    // For ease-of-use and performance, use GetValue().
+    // It may return the raw internal value or the last converted value instead of deserializing.
+    str = node.GetValue<string>();
+```
+
+### Cast operators
+No explicit or implicit cast operators exist since `GetValue<T>` allows for a single, consistent programming model. This also works on not just the "known" types in the CLR but also custom data types known by the serializer.
+
+Cast operators do not work in all languages such as F# and even for C# would not work for all types:
+```cs
+JsonArray jArray = ...
+jArray[0] = "hello"; // Possible though a string implicit operator.
+jArray[1] = myCustomDataType; // We can't know all types, so this won't work.
```

Comment:
There seems to be some magic happening here. Do we really need to support this? I am guessing some hidden serialization is happening under the covers? Maybe let the user write the few methods calls explicitly, themselves.

Reply:
None of the cast operators are required. The `JsonValue<T>` constructor can always be used to create a value, and the `value.GetValue<T>()` method can be used to obtain the value.

Assuming we want an implicit cast operator to initialize a value (safely), then I think an explicit cast operator makes sense for obtaining the value (not so safe; may throw).